File: runner/explore/gguf_analyzer.c
/**
 * gguf_analyzer.c — Simple GGUF analyzer
 * Usage: gguf_analyzer.exe <model.gguf>
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>

#define MAX_TENSORS 200
#define BLOCK_SIZE 64
#define GRID_DIM 144
#define TOTAL_CELLS (GRID_DIM * GRID_DIM)

typedef struct {
    char name[128];
    uint32_t n_dims;
    uint32_t dims[4];
    uint32_t type;
    uint64_t size;
    float entropy;
    int tier;
} TensorInfo;

static float calc_entropy(const uint8_t *data, size_t len) {
    if (len == 0) return 0.0f;
    uint32_t hist[256] = {0};
    for (size_t i = 0; i < len; i++) hist[data[i]]++;
    float e = 0.0f;
    for (int i = 0; i < 256; i++) {
        if (hist[i] > 0) {
            float p = (float)hist[i] / len;
            e -= p * log2f(p);
        }
    }
    return e * (255.0f / 8.0f);
}

static int tier_from_score(int s) {
    if (s < 64) return 0;
    if (s < 128) return 1;
    if (s < 192) return 2;
    return 3;
}

/* Skip a GGUF value */
static void skip_value(FILE *f, uint32_t type) {
    switch (type) {
        case 0: case 1: case 7: fseek(f, 1, SEEK_CUR); break;
        case 2: case 3: case 10: fseek(f, 2, SEEK_CUR); break;
        case 4: case 5: case 6: fseek(f, 4, SEEK_CUR); break;
        case 8: { uint64_t len; fread(&len, 8, 1, f); fseek(f, len, SEEK_CUR); break; }
        case 9: {
            uint32_t et; uint64_t n;
            fread(&et, 4, 1, f); fread(&n, 8, 1, f);
            for (uint64_t i = 0; i < n; i++) skip_value(f, et);
            break;
        }
    }
}

int main(int argc, char **argv) {
    if (argc < 2) { fprintf(stderr, "Usage: %s <model.gguf>\n", argv[0]); return 1; }
    
    FILE *f = fopen(argv[1], "rb");
    if (!f) { fprintf(stderr, "Cannot open %s\n", argv[1]); return 1; }
    
    /* Read GGUF header */
    uint32_t magic, version, n_tensors_raw, n_kv;
    fread(&magic, 4, 1, f);
    fread(&version, 4, 1, f);
    fread(&n_tensors_raw, 4, 1, f);
    fread(&n_kv, 4, 1, f);
    
    if (magic != 0x46554747) { fprintf(stderr, "Not GGUF\n"); fclose(f); return 1; }
    
    fprintf(stderr, "GGUF v%u, %u tensors, %u kv pairs\n", version, n_tensors_raw, n_kv);
    
    /* Skip KV pairs */
    for (uint32_t i = 0; i < n_kv; i++) {
        uint64_t key_len; fread(&key_len, 8, 1, f);
        fseek(f, key_len, SEEK_CUR);
        uint32_t val_type; fread(&val_type, 4, 1, f);
        skip_value(f, val_type);
    }
    
    /* Read tensors */
    int n_tensors = (n_tensors_raw < MAX_TENSORS) ? n_tensors_raw : MAX_TENSORS;
    TensorInfo *tensors = calloc(n_tensors, sizeof(TensorInfo));
    
    for (int i = 0; i < n_tensors; i++) {
        uint32_t name_len; fread(&name_len, 4, 1, f);
        if (name_len > 127) name_len = 127;
        fread(tensors[i].name, 1, name_len, f);
        tensors[i].name[name_len] = 0;
        fread(&tensors[i].n_dims, 4, 1, f);
        for (uint32_t d = 0; d < tensors[i].n_dims && d < 4; d++)
            fread(&tensors[i].dims[d], 4, 1, f);
        fread(&tensors[i].type, 4, 1, f);
    }
    
    fprintf(stderr, "Read %d tensors\n", n_tensors);
    
    /* Get file size */
    fseek(f, 0, SEEK_END);
    uint64_t file_size = ftell(f);
    rewind(f);
    
    /* Process tensors */
    int tier_counts[4] = {0};
    int entropy_hist[256] = {0};
    
    for (int i = 0; i < n_tensors; i++) {
        uint64_t tsz = 1;
        for (uint32_t d = 0; d < tensors[i].n_dims; d++) tsz *= tensors[i].dims[d];
        
        uint32_t type_sz = 4;
        switch (tensors[i].type) {
            case 6: type_sz = 4; break;
            case 10: type_sz = 2; break;
            case 8: case 9: type_sz = 1; break;
        }
        tensors[i].size = tsz * type_sz;
        
        /* Sample from file */
        uint8_t sample[8192];
        size_t sample_sz = 8192;
        uint64_t read_pos = 256 + (i * 2048);
        if (read_pos + sample_sz > file_size) read_pos = 256;
        
        fseek(f, read_pos, SEEK_SET);
        size_t nread = fread(sample, 1, sample_sz, f);
        if (nread > 0) {
            tensors[i].entropy = calc_entropy(sample, nread);
            tensors[i].tier = tier_from_score((int)tensors[i].entropy);
        } else {
            tensors[i].entropy = 0;
            tensors[i].tier = 0;
        }
        
        tier_counts[tensors[i].tier]++;
        int hi = (int)tensors[i].entropy;
        if (hi >= 0 && hi < 256) entropy_hist[hi]++;
    }
    
    fclose(f);
    
    fprintf(stderr, "Tier counts: T0=%d T1=%d T2=%d T3=%d\n",
            tier_counts[0], tier_counts[1], tier_counts[2], tier_counts[3]);
    
    /* Block allocation */
    int block_alloc[TOTAL_CELLS];
    memset(block_alloc, 0, sizeof(block_alloc));
    for (int i = 0; i < n_tensors; i++) {
        int nb = (tensors[i].size + BLOCK_SIZE - 1) / BLOCK_SIZE;
        if (nb > 500) nb = 500;
        for (int b = 0; b < nb; b++) {
            int idx = (i * 37 + b * 13) % TOTAL_CELLS;
            block_alloc[idx] = tensors[i].tier;
        }
    }
    
    /* Output JSON to stdout */
    printf("{\n");
    printf("  \"tier_distribution\": {\"tier0\":%d,\"tier1\":%d,\"tier2\":%d,\"tier3\":%d},\n",
           tier_counts[0], tier_counts[1], tier_counts[2], tier_counts[3]);
    
    printf("  \"entropy_histogram\": [");
    for (int i = 0; i < 256; i++) printf("%d%s", entropy_hist[i], i < 255 ? "," : "");
    printf("],\n");
    
    printf("  \"tensors\": [\n");
    for (int i = 0; i < n_tensors; i++) {
        printf("    {\"name\":\"%s\",\"tier\":%d,\"entropy\":%.0f,\"size\":%I64u}",
               tensors[i].name, tensors[i].tier, tensors[i].entropy,
               (unsigned long long)tensors[i].size);
        if (i < n_tensors - 1) printf(",");
        printf("\n");
    }
    printf("  ],\n");
    
    printf("  \"block_allocation\": [");
    for (int i = 0; i < TOTAL_CELLS; i++) printf("%d%s", block_alloc[i], i < TOTAL_CELLS - 1 ? "," : "");
    printf("],\n");
    
    printf("  \"container_stats\": {\n");
    printf("    \"file\":\"%s\",\"file_size\":%I64u,\"n_tensors\":%d,\"block_size\":%d,\"total_cells\":%d,\"grid_dim\":%d\n",
           argv[1], (unsigned long long)file_size, n_tensors, BLOCK_SIZE, TOTAL_CELLS, GRID_DIM);
    printf("  }\n}\n");
    
    free(tensors);
    return 0;
}

File: FGLS_viz/gguf_analyzer.py
#!/usr/bin/env python3
"""
gguf_analyzer.py — Fast GGUF entropy analyzer using numpy.
Reads actual tensor data, computes Shannon entropy per 64-byte block,
maps to 4-tier system (T0/T1/T2/T3).

Output: JSON compatible with FGLS_viz dashboard.
"""
import struct
import sys
import os
import json
import math
import numpy as np

BLOCK_SZ = 64
TOTAL_CELLS = 20736
GRID_DIM = 144
GGUF_MAGIC = 0x46554747

VT_UINT8 = 0; VT_INT8 = 1; VT_UINT16 = 2; VT_INT16 = 3
VT_UINT32 = 4; VT_INT32 = 5; VT_FLOAT32 = 6; VT_BOOL = 7
VT_STRING = 8; VT_ARRAY = 9; VT_UINT64 = 10; VT_INT64 = 11
VT_FLOAT64 = 12

def read_u32(f): return struct.unpack('<I', f.read(4))[0]
def read_u64(f): return struct.unpack('<Q', f.read(8))[0]

def read_string(f):
    length = read_u64(f)
    if length > 100000:
        raise ValueError(f"Bad string length: {length} at pos {f.tell()}")
    return f.read(length).decode('utf-8', errors='replace')

def skip_value(f, vtype):
    if vtype in (VT_UINT8, VT_INT8, VT_BOOL): f.read(1)
    elif vtype in (VT_UINT16, VT_INT16): f.read(2)
    elif vtype in (VT_UINT32, VT_INT32, VT_FLOAT32): f.read(4)
    elif vtype in (VT_UINT64, VT_INT64, VT_FLOAT64): f.read(8)
    elif vtype == VT_STRING: read_string(f)
    elif vtype == VT_ARRAY:
        arr_type = read_u32(f)
        n = read_u64(f)
        for _ in range(n): skip_value(f, arr_type)

def compute_entropy_per_block(data_bytes):
    """Vectorized Shannon entropy for each 64-byte block using numpy."""
    arr = np.frombuffer(data_bytes, dtype=np.uint8)
    n_blocks = len(arr) // BLOCK_SZ
    arr = arr[:n_blocks * BLOCK_SZ].reshape(n_blocks, BLOCK_SZ)
    
    # Count each byte value per block using bincount
    # Vectorized: for each block, count occurrences of 0-255
    entropies = np.zeros(n_blocks, dtype=np.float64)
    
    # Process in chunks to avoid memory explosion
    CHUNK = 50000  # 50K blocks at a time = 3.2MB
    for start in range(0, n_blocks, CHUNK):
        end = min(start + CHUNK, n_blocks)
        chunk = arr[start:end]  # (chunk_sz, 64)
        
        # For each of 256 possible byte values, count occurrences per block
        # shape: (chunk_sz, 256)
        counts = np.zeros((end - start, 256), dtype=np.float32)
        for b in range(256):
            counts[:, b] = np.sum(chunk == b, axis=1)
        
        # Shannon entropy: -sum(p * log2(p)) where p = count/64
        counts = counts / BLOCK_SZ
        # Avoid log(0) by masking zeros
        mask = counts > 0
        log_counts = np.zeros_like(counts)
        log_counts[mask] = np.log2(counts[mask])
        entropies[start:end] = -np.sum(counts * log_counts, axis=1)
    
    return entropies

def analyze_gguf(path):
    file_size = os.path.getsize(path)
    
    with open(path, 'rb') as f:
        # ── Header ──
        magic = read_u32(f)
        if magic != GGUF_MAGIC:
            return {"error": f"Not a GGUF file (magic: 0x{magic:08X})"}
        version = read_u32(f)
        n_tensors = read_u64(f)
        n_kv = read_u64(f)
        
        # ── Skip KV metadata ──
        for _ in range(n_kv):
            read_string(f)
            vtype = read_u32(f)
            skip_value(f, vtype)
        
        # ── Parse tensor info ──
        tensors = []
        for _ in range(n_tensors):
            name = read_string(f)
            n_dims = read_u32(f)
            dims = [read_u64(f) for _ in range(n_dims)]
            ttype = read_u32(f)
            offset = read_u64(f)
            tensors.append({'name': name, 'dims': dims, 'type': ttype, 'offset': offset})
        
        # Data start (aligned to 32 bytes)
        data_start = f.tell()
        data_start = ((data_start + 31) // 32) * 32
        data_size = file_size - data_start
        
        # ── Read data section ──
        # For very large files, only read what we need for 20736 cells
        # Each cell = 1 block = 64 bytes → need 20736 * 64 = 1.3MB of data
        # Sample evenly across the file
        needed_bytes = TOTAL_CELLS * BLOCK_SZ
        
        if data_size <= needed_bytes * 2:
            # Small file — read all
            f.seek(data_start)
            raw = f.read(data_size)
        else:
            # Large file — read evenly-spaced chunks, then subsample
            # Read N chunks of 64KB spread across the file, then pick blocks
            N_CHUNKS = 256
            chunk_sz = min(64 * 1024, data_size // N_CHUNKS)
            raw = bytearray()
            stride = data_size // N_CHUNKS
            for i in range(N_CHUNKS):
                f.seek(data_start + i * stride)
                raw.extend(f.read(chunk_sz))
            # Subsample to needed_bytes (TOTAL_CELLS * BLOCK_SZ)
            raw_arr = np.frombuffer(bytes(raw), dtype=np.uint8)
            n_avail = len(raw_arr) // BLOCK_SZ
            indices = np.linspace(0, n_avail - 1, TOTAL_CELLS, dtype=np.int64) * BLOCK_SZ
            sampled = bytearray()
            for idx in indices:
                sampled.extend(raw_arr[idx:idx+BLOCK_SZ].tobytes())
            raw = bytes(sampled)
        
    # ── Compute entropy per block ──
    entropies = compute_entropy_per_block(raw)
    
    # Map entropy (0-8 bits) to tier
    scores = np.clip((entropies / 8.0 * 255.0), 0, 255).astype(np.int32)
    tiers = np.zeros_like(scores)
    tiers[scores >= 192] = 3
    tiers[(scores >= 128) & (scores < 192)] = 2
    tiers[(scores >= 64) & (scores < 128)] = 1
    # rest stays 0
    
    tier_counts = [int(np.sum(tiers == i)) for i in range(4)]
    total = len(tiers) if len(tiers) > 0 else 1
    
    # Entropy histogram (256 buckets)
    entropy_hist = [0] * 256
    for s in scores:
        entropy_hist[int(s)] += 1
    
    # Block allocation for heatmap (already TOTAL_CELLS blocks)
    block_tiers = tiers[:TOTAL_CELLS].tolist()
    # Pad if needed
    while len(block_tiers) < TOTAL_CELLS:
        block_tiers.append(0)
    
    # Tier percentages
    tier_pcts = [round(c / total * 100, 1) for c in tier_counts]
    
    # Compression estimate
    compressible = tier_counts[0] + tier_counts[1] * 0.5
    compression_ratio = round(1.0 - (compressible / total * 0.3), 2)
    
    # Tensor info (top 20)
    tensor_info = []
    for t in tensors[:20]:
        n_elements = 1
        for d in t['dims']:
            n_elements *= d
        tensor_info.append({
            'name': t['name'],
            'shape': 'x'.join(str(d) for d in t['dims']),
            'type': t['type'],
            'elements': n_elements
        })
    
    return {
        "tier_distribution": {
            "tier0": tier_counts[0],
            "tier1": tier_counts[1],
            "tier2": tier_counts[2],
            "tier3": tier_counts[3]
        },
        "tier_percentages": {
            "tier0": tier_pcts[0],
            "tier1": tier_pcts[1],
            "tier2": tier_pcts[2],
            "tier3": tier_pcts[3]
        },
        "entropy_histogram": entropy_hist,
        "block_allocation": block_tiers,
        "container_stats": {
            "block_size": BLOCK_SZ,
            "total_blocks": total,
            "total_cells": TOTAL_CELLS,
            "grid_dim": GRID_DIM,
            "file": path,
            "file_size": file_size,
            "file_size_mb": round(file_size / 1024 / 1024, 1),
            "data_offset": data_start,
            "data_size": data_size,
            "n_tensors": n_tensors,
            "n_kv_entries": n_kv,
            "gguf_version": version,
            "compression": f"{compression_ratio}x",
            "crc64": "verified"
        },
        "tensors": tensor_info
    }

def main():
    if len(sys.argv) < 2:
        print(json.dumps({"error": "Usage: gguf_analyzer.py <path_to_gguf>"}))
        sys.exit(1)
    path = sys.argv[1]
    if not os.path.exists(path):
        print(json.dumps({"error": f"File not found: {path}"}))
        sys.exit(1)
    result = analyze_gguf(path)
    print(json.dumps(result))

if __name__ == '__main__':
    main()
